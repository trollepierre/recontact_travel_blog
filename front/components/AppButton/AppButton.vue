<template>
  <component
    :is="tag"
    type="button"
    class="button"
    :to="to"
    @click="(e) => allowMultipleClick ? onClick(e) : undefined"
    @click.prevent.once="(e) => allowMultipleClick ? undefined : onClick(e)">
    {{ text }}
  </component>
</template>
<script>
  export default {
    name: 'AppButton',
    props: {
      text: { type: String, default: () => '' },
      isLink: { type: Boolean, default: () => false },
      to: { type: String, default: () => undefined },
      tag: { type: String, default: () => 'button' },
      allowMultipleClick: { type: Boolean, default: () => false },
    },
    methods: {
      onClick(e) {
        e.preventDefault()
        this.$emit('click')
      },
    },
  }
</script>

<style lang="scss" scoped>
.button {
  line-height: 28px;
  color: $button-color;
  text-decoration: unset;
  font-size: 11px;
  font-family: serif;
  text-transform: uppercase;
  background: $button-bg;
  border: 1px solid $button-color;
  cursor: pointer;
  padding: 3px 5px;
  border-radius: 4px;
  width: 100%;
  font-weight: 700;
}

.button:hover {
  background: $button-hover;
  color: $button-bg;
}

</style>
