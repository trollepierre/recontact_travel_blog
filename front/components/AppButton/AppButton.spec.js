import AppButton from './AppButton.vue'

describe('Component | AppButton.vue', () => {
  let localVue
  let wrapper

  beforeEach(() => {
    localVue = createLocalVue()
  })

  describe('template', () => {
    it('should match snapshot', () => {
      wrapper = shallowMount(AppButton, { localVue })

      expect(wrapper).toMatchSnapshot()
    })

    it('should display anchor when tag is a', () => {
      const propsData = { tag: 'a', href: 'http://rec.me', text: 'anchor' }

      wrapper = shallowMount(AppButton, { localVue, propsData })

      expect(wrapper).toMatchSnapshot()
    })

    it('should display nuxt link when tag is nuxt link', () => {
      const consoleError = console.error
      console.error = jest.fn()
      const propsData = { tag: 'nuxt-link', to: 'http://rec.me', text: 'link' }

      wrapper = shallowMount(AppButton, { localVue, propsData })

      expect(wrapper).toMatchSnapshot()
      console.error = consoleError
    })

    it('should hide component when needed on app-header', () => {
      const propsData = { hide: true }

      wrapper = shallowMount(AppButton, { localVue, propsData })

      expect(wrapper.find('button').classes()).toContain('hidden')
    })
  })

  describe('methods', () => {
    it('should handle click - and prevent multiple click', () => {
      // Given
      wrapper = shallowMount(AppButton, { localVue })

      // When
      wrapper.find('button').trigger('click')
      wrapper.find('button').trigger('click')

      // Then
      expect(wrapper).toEmit('click')
      expect(wrapper.emitted().click).toHaveLength(1)
    })

    it('should handle multiple click', () => {
      // Given
      const propsData = { allowMultipleClick: true }
      wrapper = shallowMount(AppButton, { localVue, propsData })

      // When
      wrapper.find('button').trigger('click')
      wrapper.find('button').trigger('click')

      // Then
      expect(wrapper.emitted().click).toHaveLength(2)
    })
  })
})
