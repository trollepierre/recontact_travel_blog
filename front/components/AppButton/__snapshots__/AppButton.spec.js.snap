// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component | AppButton.vue template should display anchor when tag is a 1`] = `
<a class="button" href="http://rec.me">
  anchor
</a>
`;

exports[`Component | AppButton.vue template should display nuxt link when tag is nuxt link 1`] = `
<nuxt-link to="http://rec.me" class="button">
  link
</nuxt-link>
`;

exports[`Component | AppButton.vue template should match snapshot 1`] = `
<button type="button" class="button">

</button>
`;
